exp/api/v1alpha3/awsmanagedcontrolplane_types.go

```diff
@@ -0,0 +1,184 @@
+)
+
+// AWSManagedControlPlaneSpec defines the desired state of AWSManagedControlPlane
+type AWSManagedControlPlaneSpec struct {
+	// NetworkSpec encapsulates all things related to AWS network.
+	NetworkSpec infrav1.NetworkSpec `json:"networkSpec,omitempty"`
+
+	// The AWS Region the cluster lives in.
+	Region string `json:"region,omitempty"`
+
+	// SSHKeyName is the name of the ssh key to attach to the bastion host. Valid values are empty string (do not use SSH keys), a valid SSH key name, or omitted (use the default SSH key name)
+	// +optional
+	SSHKeyName *string `json:"sshKeyName,omitempty"`
+
+	// Version defines the desired Kubernetes version. If no version number
+	// is supplied then the latest version of Kubernetesthat EKS supports
+	// will be used.
+	// +kubebuilder:validation:MinLength:=2
+	// +kubebuilder:validation:Pattern:=^v(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.?$
+	// +optional
+	Version *string `json:"version"`
+
+	// RoleName specifies the name of IAM role that gives EKS
+	// permission to make API calls. If no name is supplied
+	// then a role is created. If the role is pre-existing
+	// we will treat it as unmanaged and not delete it on
+	// deletion
+	// +kubebuilder:validation:MinLength:=2
+	// +optional
+	RoleName *string `json:"roleName"`
+
+	// RoleAdditionalPolicies allows you to attach additional polices to
+	// the control plane role.
+	// +optional
+	RoleAdditionalPolicies *[]string `json:"roleAdditionalPolicies"`
+
+	// Logging specifies which EKS Cluster logs should be enabled. Entries for
+	// each of the enabled logs will be sent to CloudWatch
+	// +optional
+	Logging *ControlPlaneLoggingSpec `json:"logging,omitempty"`
+
+	// EncryptionConfig specifies the encryption configuration for the cluster
+	// +optional
+	EncryptionConfig *EncryptionConfig `json:"encryptionConfig,omitempty"`
+
+	// AdditionalTags is an optional set of tags to add to AWS resources managed by the AWS provider, in addition to the
+	// ones added by default.
+	// +optional
+	AdditionalTags infrav1.Tags `json:"additionalTags,omitempty"`
+
+	// Endpoints specifies access to this cluster's control plane endpoints
+	// +optional
+	EndpointAccess EndpointAccess `json:"endpointAccess,omitempty"`
+
+	// ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
+
+	// Bastion contains options to configure the bastion host.
+	// +optional
+	Bastion infrav1.Bastion `json:"bastion"`
+
+	// TokenMethod is used to specify the method for obtaining a client token for communicating with EKS
+	// iam-authenticator - obtains a client token using iam-authentictor
+	// aws-cli - obtains a client token using the AWS CLI
+	// Defaults to iam-authenticator
+	// +kubebuilder:default=iam-authenticator
+	// +kubebuilder:validation:Enum=iam-authenticator;aws-cli
+	TokenMethod *EKSTokenMethod `json:"tokenMethod,omitempty"`
```

Comment:
This totally slipped my mind when we were initially discussing the kubeconfig generation - CAPI uses the kubeconfig, not CAPA, so in order for this to work either the aws CLI or the iam authenticator needs to be bundled in the CAPI image and then it also needs to mount AWS credentials. 

It may be worth exploring creating a CAPI serviceaccount and generating the kubeconfig from that. Right now, CAPI is caught in a loop: `getting credentials: exec: exec: \"aws\": executable file not found in $PATH" `

Reply:
The generated kubeconfig works fine for local use, though, save for the base64 issue i mentioned in another comment.

Reply:
I've had a nagging doubt around the kubeconfig for a while and thats great @rudoi that you've been taking things for a spin. 

@rudoi @detiber - so the CAPI controllers would have a service account that they can use when using the generated kubeconfig?

Reply:
I think this can work if the generated kubeconfig secret contains a kubeconfig created from a serviceaccount token. At that point CAPI would be able to reconcile Cluster objects properly. then we slap a big warning on the docs like "hey, for AWSManagedClusters CAPI uses a serviceaccount, use the `aws` CLI to get credentials.  

A separate issue here (and I think this is fine for a follow-up PR) - By default in EKS, the only IAM entity that can access the Kubernetes cluster is the entity that created it. It would be awesome if the spec for AWSManagedControlPlane had a way to specify additional IAM roles (or at least one for an admin role).

Reply:
Yes that right, by default on creation only the user that created the cluster has access.  That would be great if we add `AdditionalRoles` or something in the control plane spec.  That should be too hard at all, especially with: https://github.com/kubernetes-sigs/aws-iam-authenticator#full-configuration-format.

And also agreed on the docs update. Shall we create a issues to cover these and link to the parent EKS issue?

Reply:
Let's do that - i'll create issues for the AdditionalRoles and docs updates.

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1850

Reply:
there's sample python code here https://github.com/kubernetes-sigs/aws-iam-authenticator#api-authorization-from-outside-a-cluster .

I think it should be possible to use the AWS SDK for Go to generate the presigned URL against the STS get caller identity API call, and pass that in as the initial bearer token to then generate a service account and store the kubeconfig back on the management server.

Reply:
Thats a great idea @randomvariable. I will try that out now.